pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/impl/internal/CognitiveComplexityVisitor.java

```diff
@@ -249,7 +249,7 @@ public Object visit(ASTBlockStatement node, Object data) {
     @Override
     public Object visit(ASTMethodDeclaration node, Object data) {
         State state = (State) data;
-        state.setMethodName(node.getName());
+        state.setMethodName(node.getQualifiedName().toString());
```

Comment:
Qualified names are deprecated, but you could just store the `node` into the state here ...

Reply:
Oh I see. Btw what would a proper check for recursion? Do I understand correctly that we consider any call to a method in the stack as recursive? For example if we call `run` from `run` and we call `foo` from `run` - both are recursive calls, right?

Reply:
I would say yes, any member of the stack counts... If you compare the above code to how it would be written with a lambda, this is consistent: `((Runnable) () -> foo()).run()` has the same complexity as the `new Runnable() {...}` above. If we just considered the top one, then the lambda would have *higher* complexity. OTOH, I'm sure this is not perfect and there are counter arguments... This is worth documenting with a comment in the implementation

Reply:
Re (1) makes sense, but I would make it rather an option so the default configuration matches the SonarSource spec.

Reply:
Re docs - could you pls add it? I am not 100% what exactly you want to write.